Support main methods from traits in Scala 3.4 (200USD Bounty)

---

**From the maintainer Li Haoyi**: I'm putting a 200USD bounty on this issue, payable by bank transfer on a merged PR resolving this issue, including a unit test to avoid regression. The issue is straightforward to reproduce.

---

My Main object has grown too large and I want to split it. I tried to create a trait for each entry point and then making the Main object extends all those traits. But in such case, mainargs does not find any of the @main annotated entry points in any of the traits, while it worked flawlessly when all those methods were together in the same Main object.

```scala
trait CommandList {
  @main
  def list(@arg v: String): Unit = ???
}

trait CommandCopy {
  @main
  def copy(@arg from: String, @arg to: String): Unit = ???
}

object Main extends CommandList with CommandCopy {
  def main(args: Array[String]): Unit = ParserForMethods(this).runOrExit(args)
}
```

Is there any way to support this use case? Without having to duplicate the entry point in all of them, of course.

For an example like this one:
```scala
trait CommandCopy {
  @main
  def copy(@arg from: String, @arg to: String): Unit = ???
}

class Joined extends  CommandCopy {
  @main
  def test(@arg from: String, @arg to: String): Unit = ???

  val parser = ParserForMethods(this)
}
```

The issue manifests itself as a compiler crash:
```
[error] ## Exception when compiling 23 sources to /home/aleksander/repos/mainargs/out/mainargs/jvm/3.3.1/test/compile.dest/classes
[error] java.lang.AssertionError: assertion failed: missing outer accessor in class Joined
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.transform.ExplicitOuter$.dotty$tools$dotc$transform$ExplicitOuter$$$outerParamAccessor(ExplicitOuter.scala:237)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.loop$1(ExplicitOuter.scala:474)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.path$extension(ExplicitOuter.scala:483)
```

The code that gets generated looks like this:
```scala
(new ParserForMethods[Joined](
  MethodMains.apply[Joined](
    List.apply[MainData[Any, Joined]](
      MainData.create[Any, Joined](
        "copy",
        new main(main.$lessinit$greater$default$1, main.$lessinit$greater$default$2),
        List.apply[ArgSig](
          ArgSig.create[String, Joined](
            "from",
            new arg(arg.$lessinit$greater$default$1, arg.$lessinit$greater$default$2, arg.$lessinit$greater$default$3, arg.$lessinit$greater$default$4, arg.$lessinit$greater$default$5, arg.$lessinit$greater$default$6),
            None
          )(StringRead), 
          ArgSig.create[String, Joined](
            "to",
            new arg(arg.$lessinit$greater$default$1, arg.$lessinit$greater$default$2, arg.$lessinit$greater$default$3, arg.$lessinit$greater$default$4, arg.$lessinit$greater$default$5, arg.$lessinit$greater$default$6),
            None
          )(StringRead)),
        ((b: Joined, params: [A >: Nothing <: Any] =>> Seq[A][Any]) =>
          CommandCopy.this.copy(
            Seq.apply[Seq[Any]](params).apply(0).apply(0).asInstanceOf[String],
            Seq.apply[Seq[Any]](params).apply(0).apply(1).asInstanceOf[String]
          )
        )),
      MainData.create[Any, Joined](
        "test",
        new main(main.$lessinit$greater$default$1, main.$lessinit$greater$default$2),
        List.apply[ArgSig](
          ArgSig.create[String, Joined](
            "from",
            new arg(arg.$lessinit$greater$default$1, arg.$lessinit$greater$default$2, arg.$lessinit$greater$default$3, arg.$lessinit$greater$default$4, arg.$lessinit$greater$default$5, arg.$lessinit$greater$default$6),
            None
          )(StringRead),
        ArgSig.create[String, Joined](
          "to",
          new arg(arg.$lessinit$greater$default$1, arg.$lessinit$greater$default$2, arg.$lessinit$greater$default$3, arg.$lessinit$greater$default$4, arg.$lessinit$greater$default$5, arg.$lessinit$greater$default$6),
          None
        )(StringRead)),
        ((`b₂`: Joined, `params₂`: [A >: Nothing <: Any] =>> Seq[A][Any]) =>
          Joined.this.test(
            Seq.apply[Seq[Any]](`params₂`).apply(0).apply(0).asInstanceOf[String],
            Seq.apply[Seq[Any]](`params₂`).apply(0).apply(1).asInstanceOf[String]
          )
        ))), (() => Joined.this))
    ): ParserForMethods[Joined])
```

The issue itself stems from this:
```scala
((b: Joined, params: [A >: Nothing <: Any] =>> Seq[A][Any]) =>
          CommandCopy.this.copy(
            Seq.apply[Seq[Any]](params).apply(0).apply(0).asInstanceOf[String],
            Seq.apply[Seq[Any]](params).apply(0).apply(1).asInstanceOf[String]
          )
        )),
```
...so my bet is that this shouldn't be `CommandCopy.this.copy(...)` but `b.copy(...)`

--- a/mainargs/src-3/Macros.scala
+++ b/mainargs/src-3/Macros.scala
@@ -11,8 +11,9 @@ object Macros {
     val annotatedMethodsWithMainAnnotations = allMethods.flatMap { methodSymbol =>
       methodSymbol.getAnnotation(mainAnnotation).map(methodSymbol -> _)
     }.sortBy(_._1.pos.map(_.start))
-    val mainDatas = Expr.ofList(annotatedMethodsWithMainAnnotations.map { (annotatedMethod, mainAnnotationInstance) =>
-      createMainData[Any, B](annotatedMethod, mainAnnotationInstance)
+    val mainDatas = Expr.ofList(annotatedMethodsWithMainAnnotations.map {
+      (annotatedMethod, mainAnnotationInstance) =>
+        createMainData[Any, B](annotatedMethod, mainAnnotationInstance)
     })
 
     '{
@@ -26,7 +27,7 @@ object Macros {
     import quotes.reflect._
     val typeReprOfB = TypeRepr.of[B]
     val companionModule = typeReprOfB match {
-      case TypeRef(a,b) => TermRef(a,b)
+      case TypeRef(a, b) => TermRef(a, b)
     }
     val typeSymbolOfB = typeReprOfB.typeSymbol
     val companionModuleType = typeSymbolOfB.companionModule.tree.asInstanceOf[ValDef].tpt.tpe.asType
@@ -51,31 +52,36 @@ object Macros {
         '{ new ParserForClass[B](${ mainData }, () => ${ Ident(companionModule).asExpr }) }
   }
 
-  def createMainData[T: Type, B: Type](using Quotes)
-                                      (method: quotes.reflect.Symbol,
-                                       mainAnnotation: quotes.reflect.Term): Expr[MainData[T, B]] = {
+  def createMainData[T: Type, B: Type](using
+      Quotes
+  )(method: quotes.reflect.Symbol, mainAnnotation: quotes.reflect.Term): Expr[MainData[T, B]] = {
     createMainData[T, B](method, mainAnnotation, method.paramSymss)
   }
 
-  def createMainData[T: Type, B: Type](using Quotes)
-                                      (method: quotes.reflect.Symbol,
-                                       mainAnnotation: quotes.reflect.Term,
-                                       annotatedParamsLists: List[List[quotes.reflect.Symbol]]): Expr[MainData[T, B]] = {
+  def createMainData[T: Type, B: Type](using Quotes)(
+      method: quotes.reflect.Symbol,
+      mainAnnotation: quotes.reflect.Term,
+      annotatedParamsLists: List[List[quotes.reflect.Symbol]]
+  ): Expr[MainData[T, B]] = {
 
     import quotes.reflect.*
-    val params = method.paramSymss.headOption.getOrElse(report.throwError("Multiple parameter lists not supported"))
+    val params = method.paramSymss.headOption.getOrElse(
+      report.throwError("Multiple parameter lists not supported")
+    )
     val defaultParams = getDefaultParams(method)
     val argSigsExprs = params.zip(annotatedParamsLists.flatten).map { paramAndAnnotParam =>
       val param = paramAndAnnotParam._1
       val annotParam = paramAndAnnotParam._2
       val paramTree = param.tree.asInstanceOf[ValDef]
       val paramTpe = paramTree.tpt.tpe
-      val arg = annotParam.getAnnotation(argAnnotation).map(_.asExprOf[mainargs.arg]).getOrElse('{ new mainargs.arg() })
+      val arg = annotParam.getAnnotation(argAnnotation).map(_.asExprOf[mainargs.arg]).getOrElse('{
+        new mainargs.arg()
+      })
       val paramType = paramTpe.asType
       paramType match
         case '[t] =>
           val defaultParam: Expr[Option[B => t]] = defaultParams.get(param) match {
-            case Some('{ $v: `t`}) => '{ Some(((_: B) => $v)) }
+            case Some('{ $v: `t` }) => '{ Some(((_: B) => $v)) }
             case None => '{ None }
           }
           val tokensReader = Expr.summon[mainargs.TokensReader[t]].getOrElse {
@@ -84,39 +90,53 @@ object Macros {
               param.pos.get
             )
           }
-          '{ (ArgSig.create[t, B](${ Expr(param.name) }, ${ arg }, ${ defaultParam })(using ${ tokensReader })) }
+          '{
+            (ArgSig.create[t, B](${ Expr(param.name) }, ${ arg }, ${ defaultParam })(using
+            ${ tokensReader }))
+          }
     }
     val argSigs = Expr.ofList(argSigsExprs)
 
     val invokeRaw: Expr[(B, Seq[Any]) => T] = {
-      def callOf(args: Expr[Seq[Any]]) = call(method, '{ Seq( ${ args }) }).asExprOf[T]
-      '{ ((b: B, params: Seq[Any]) => ${ callOf('{ params }) }) }
+
+      def callOf(methodOwner: Expr[Any], args: Expr[Seq[Any]]) =
+        call(methodOwner, method, '{ Seq($args) }).asExprOf[T]
+
+      '{ (b: B, params: Seq[Any]) => ${ callOf('b, 'params) } }
     }
-    '{ MainData.create[T, B](${ Expr(method.name) }, ${ mainAnnotation.asExprOf[mainargs.main] }, ${ argSigs }, ${ invokeRaw }) }
+    '{
+      MainData.create[T, B](
+        ${ Expr(method.name) },
+        ${ mainAnnotation.asExprOf[mainargs.main] },
+        $argSigs,
+        $invokeRaw
+      )
+    }
   }
 
-  /** Call a method given by its symbol.
-    *
-    * E.g.
-    *
-    * assuming:
-    *
-    *   def foo(x: Int, y: String)(z: Int)
-    *
-    *   val argss: List[List[Any]] = ???
-    *
-    * then:
-    *
-    *   call(<symbol of foo>, '{argss})
-    *
-    * will expand to:
-    *
-    *   foo(argss(0)(0), argss(0)(1))(argss(1)(0))
-    *
-    */
+  /**
+   * Call a method given by its symbol.
+   *
+   * E.g.
+   *
+   * assuming:
+   *
+   *   def foo(x: Int, y: String)(z: Int)
+   *
+   *   val argss: List[List[Any]] = ???
+   *
+   * then:
+   *
+   *   call(<symbol of foo>, '{argss})
+   *
+   * will expand to:
+   *
+   *   foo(argss(0)(0), argss(0)(1))(argss(1)(0))
+   */
   private def call(using Quotes)(
-    method: quotes.reflect.Symbol,
-    argss: Expr[Seq[Seq[Any]]]
+      methodOwner: Expr[Any],
+      method: quotes.reflect.Symbol,
+      argss: Expr[Seq[Seq[Any]]]
   ): Expr[_] = {
     // Copy pasted from Cask.
     // https://github.com/com-lihaoyi/cask/blob/65b9c8e4fd528feb71575f6e5ef7b5e2e16abbd9/cask/src-3/cask/router/Macros.scala#L106
@@ -127,25 +147,21 @@ object Macros {
       report.throwError("At least one parameter list must be declared.", method.pos.get)
     }
 
-    val fct = Ref(method)
-
     val accesses: List[List[Term]] = for (i <- paramss.indices.toList) yield {
       for (j <- paramss(i).indices.toList) yield {
         val tpe = paramss(i)(j).tree.asInstanceOf[ValDef].tpt.tpe
         tpe.asType match
-          case '[t] => '{ $argss(${Expr(i)})(${Expr(j)}).asInstanceOf[t] }.asTerm
+          case '[t] => '{ $argss(${ Expr(i) })(${ Expr(j) }).asInstanceOf[t] }.asTerm
       }
     }
 
-    val base = Apply(fct, accesses.head)
-    val application: Apply = accesses.tail.foldLeft(base)((lhs, args) => Apply(lhs, args))
-    val expr = application.asExpr
-    expr
+    methodOwner.asTerm.select(method).appliedToArgss(accesses).asExpr
   }
 
-
   /** Lookup default values for a method's parameters. */
-  private def getDefaultParams(using Quotes)(method: quotes.reflect.Symbol): Map[quotes.reflect.Symbol, Expr[Any]] = {
+  private def getDefaultParams(using
+      Quotes
+  )(method: quotes.reflect.Symbol): Map[quotes.reflect.Symbol, Expr[Any]] = {
     // Copy pasted from Cask.
     // https://github.com/com-lihaoyi/cask/blob/65b9c8e4fd528feb71575f6e5ef7b5e2e16abbd9/cask/src-3/cask/router/Macros.scala#L38
     import quotes.reflect._
@@ -158,7 +174,7 @@ object Macros {
 
     val idents = method.owner.tree.asInstanceOf[ClassDef].body
 
-    idents.foreach{
+    idents.foreach {
       case deff @ DefDef(Name(idx), _, _, _) =>
         val expr = Ref(deff.symbol).asExpr
         defaults += (params(idx.toInt - 1) -> expr)

--- a/mainargs/src-3/Macros.scala
+++ b/mainargs/src-3/Macros.scala
@@ -11,9 +11,8 @@ object Macros {
     val annotatedMethodsWithMainAnnotations = allMethods.flatMap { methodSymbol =>
       methodSymbol.getAnnotation(mainAnnotation).map(methodSymbol -> _)
     }.sortBy(_._1.pos.map(_.start))
-    val mainDatas = Expr.ofList(annotatedMethodsWithMainAnnotations.map {
-      (annotatedMethod, mainAnnotationInstance) =>
-        createMainData[Any, B](annotatedMethod, mainAnnotationInstance)
+    val mainDatas = Expr.ofList(annotatedMethodsWithMainAnnotations.map { (annotatedMethod, mainAnnotationInstance) =>
+      createMainData[Any, B](annotatedMethod, mainAnnotationInstance)
     })
 
     '{
@@ -27,7 +26,7 @@ object Macros {
     import quotes.reflect._
     val typeReprOfB = TypeRepr.of[B]
     val companionModule = typeReprOfB match {
-      case TypeRef(a, b) => TermRef(a, b)
+      case TypeRef(a,b) => TermRef(a,b)
     }
     val typeSymbolOfB = typeReprOfB.typeSymbol
     val companionModuleType = typeSymbolOfB.companionModule.tree.asInstanceOf[ValDef].tpt.tpe.asType
@@ -52,36 +51,31 @@ object Macros {
         '{ new ParserForClass[B](${ mainData }, () => ${ Ident(companionModule).asExpr }) }
   }
 
-  def createMainData[T: Type, B: Type](using
-      Quotes
-  )(method: quotes.reflect.Symbol, mainAnnotation: quotes.reflect.Term): Expr[MainData[T, B]] = {
+  def createMainData[T: Type, B: Type](using Quotes)
+                                      (method: quotes.reflect.Symbol,
+                                       mainAnnotation: quotes.reflect.Term): Expr[MainData[T, B]] = {
     createMainData[T, B](method, mainAnnotation, method.paramSymss)
   }
 
-  def createMainData[T: Type, B: Type](using Quotes)(
-      method: quotes.reflect.Symbol,
-      mainAnnotation: quotes.reflect.Term,
-      annotatedParamsLists: List[List[quotes.reflect.Symbol]]
-  ): Expr[MainData[T, B]] = {
+  def createMainData[T: Type, B: Type](using Quotes)
+                                      (method: quotes.reflect.Symbol,
+                                       mainAnnotation: quotes.reflect.Term,
+                                       annotatedParamsLists: List[List[quotes.reflect.Symbol]]): Expr[MainData[T, B]] = {
 
     import quotes.reflect.*
-    val params = method.paramSymss.headOption.getOrElse(
-      report.throwError("Multiple parameter lists not supported")
-    )
+    val params = method.paramSymss.headOption.getOrElse(report.throwError("Multiple parameter lists not supported"))
     val defaultParams = getDefaultParams(method)
     val argSigsExprs = params.zip(annotatedParamsLists.flatten).map { paramAndAnnotParam =>
       val param = paramAndAnnotParam._1
       val annotParam = paramAndAnnotParam._2
       val paramTree = param.tree.asInstanceOf[ValDef]
       val paramTpe = paramTree.tpt.tpe
-      val arg = annotParam.getAnnotation(argAnnotation).map(_.asExprOf[mainargs.arg]).getOrElse('{
-        new mainargs.arg()
-      })
+      val arg = annotParam.getAnnotation(argAnnotation).map(_.asExprOf[mainargs.arg]).getOrElse('{ new mainargs.arg() })
       val paramType = paramTpe.asType
       paramType match
         case '[t] =>
           val defaultParam: Expr[Option[B => t]] = defaultParams.get(param) match {
-            case Some('{ $v: `t` }) => '{ Some(((_: B) => $v)) }
+            case Some('{ $v: `t`}) => '{ Some(((_: B) => $v)) }
             case None => '{ None }
           }
           val tokensReader = Expr.summon[mainargs.TokensReader[t]].getOrElse {
@@ -90,10 +84,7 @@ object Macros {
               param.pos.get
             )
           }
-          '{
-            (ArgSig.create[t, B](${ Expr(param.name) }, ${ arg }, ${ defaultParam })(using
-            ${ tokensReader }))
-          }
+          '{ (ArgSig.create[t, B](${ Expr(param.name) }, ${ arg }, ${ defaultParam })(using ${ tokensReader })) }
     }
     val argSigs = Expr.ofList(argSigsExprs)
 
@@ -104,35 +95,28 @@ object Macros {
 
       '{ (b: B, params: Seq[Any]) => ${ callOf('b, 'params) } }
     }
-    '{
-      MainData.create[T, B](
-        ${ Expr(method.name) },
-        ${ mainAnnotation.asExprOf[mainargs.main] },
-        $argSigs,
-        $invokeRaw
-      )
-    }
+    '{ MainData.create[T, B](${ Expr(method.name) }, ${ mainAnnotation.asExprOf[mainargs.main] }, ${ argSigs }, ${ invokeRaw }) }
   }
 
-  /**
-   * Call a method given by its symbol.
-   *
-   * E.g.
-   *
-   * assuming:
-   *
-   *   def foo(x: Int, y: String)(z: Int)
-   *
-   *   val argss: List[List[Any]] = ???
-   *
-   * then:
-   *
-   *   call(<symbol of foo>, '{argss})
-   *
-   * will expand to:
-   *
-   *   foo(argss(0)(0), argss(0)(1))(argss(1)(0))
-   */
+  /** Call a method given by its symbol.
+    *
+    * E.g.
+    *
+    * assuming:
+    *
+    *   def foo(x: Int, y: String)(z: Int)
+    *
+    *   val argss: List[List[Any]] = ???
+    *
+    * then:
+    *
+    *   call(<symbol of foo>, '{argss})
+    *
+    * will expand to:
+    *
+    *   foo(argss(0)(0), argss(0)(1))(argss(1)(0))
+    *
+    */
   private def call(using Quotes)(
       methodOwner: Expr[Any],
       method: quotes.reflect.Symbol,
@@ -151,17 +135,16 @@ object Macros {
       for (j <- paramss(i).indices.toList) yield {
         val tpe = paramss(i)(j).tree.asInstanceOf[ValDef].tpt.tpe
         tpe.asType match
-          case '[t] => '{ $argss(${ Expr(i) })(${ Expr(j) }).asInstanceOf[t] }.asTerm
+          case '[t] => '{ $argss(${Expr(i)})(${Expr(j)}).asInstanceOf[t] }.asTerm
       }
     }
 
     methodOwner.asTerm.select(method).appliedToArgss(accesses).asExpr
   }
+    
 
   /** Lookup default values for a method's parameters. */
-  private def getDefaultParams(using
-      Quotes
-  )(method: quotes.reflect.Symbol): Map[quotes.reflect.Symbol, Expr[Any]] = {
+  private def getDefaultParams(using Quotes)(method: quotes.reflect.Symbol): Map[quotes.reflect.Symbol, Expr[Any]] = {
     // Copy pasted from Cask.
     // https://github.com/com-lihaoyi/cask/blob/65b9c8e4fd528feb71575f6e5ef7b5e2e16abbd9/cask/src-3/cask/router/Macros.scala#L38
     import quotes.reflect._
@@ -174,7 +157,7 @@ object Macros {
 
     val idents = method.owner.tree.asInstanceOf[ClassDef].body
 
-    idents.foreach {
+    idents.foreach{
       case deff @ DefDef(Name(idx), _, _, _) =>
         val expr = Ref(deff.symbol).asExpr
         defaults += (params(idx.toInt - 1) -> expr)

--- a/mainargs/test/src/MultiTraitTests.scala
+++ b/mainargs/test/src/MultiTraitTests.scala
@@ -1,0 +1,26 @@
+package mainargs
+import utest._
+
+trait CommandList {
+  @main
+  def list(@arg v: String): String = v
+}
+
+trait CommandCopy {
+  @main
+  def copy(@arg from: String, @arg to: String): (String, String) = (from, to)
+}
+
+object Joined extends CommandCopy with CommandList {
+  @main
+  def test(@arg from: String, @arg to: String): (String, String) = (from, to)
+}
+
+object MultiTraitTests extends TestSuite {
+  val check = new Checker(ParserForMethods(Joined), allowPositional = true)
+  val tests = Tests {
+    test - check(List("copy", "fromArg", "toArg"), Result.Success(("fromArg", "toArg")))
+    test - check(List("test", "fromArg", "toArg"), Result.Success(("fromArg", "toArg")))
+    test - check(List("list", "vArg"), Result.Success("vArg"))
+  }
+}

